src/web/web_sockets.cpp

```diff
@@ -0,0 +1,560 @@
+    if (it == query_work_map_.end())
+    {
+        LOG_ERROR(LOG_SERVER)
+            << "Unmatched work item by id: " << response.id();
+        return;
+    }
+
+    const auto work = it->second;
+    query_work_map_.erase(it);
+
+    data_source istream(response.data());
+    istream_reader source(istream);
+
+    const auto ec = source.read_error_code();
+
+    if (ec)
+    {
+        const std::string error = work.command +
+            " failed: " + ec.message();
+
+        send(work.connection, to_json(error, ec));
+        return;
+    }
+
+    auto handler = translator_.find(work.command);
+    if (handler == translator_.end())
+    {
+        const std::string error = work.command +
+            " is not a valid query command.";
+
+        send(work.connection, error);
+        return;
+    }
+
+    const auto payload = source.read_bytes();
+    handler->second.decode(payload, work.connection);
+}
+
+void web_sockets::remove_query_work(mg_connection_ptr connection)
+{
+    auto is_match = [&connection](const query_work_item& item)
+    {
+        return item.connection == connection;
+    };
+
+    std::remove_if(pending_query_work_list_.begin(),
+                   pending_query_work_list_.end(), is_match);
+}
+
+void web_sockets::send(mg_connection_ptr connection,
+    const std::string& data_string)
+{
+    mg_send_websocket_frame(connection, WEBSOCKET_OP_TEXT,
+        data_string.c_str(), data_string.size());
+}
+
+void web_sockets::broadcast(const std::string& data_string)
+{
+    for (auto& connection : connections_)
+        send(connection, data_string);
+}
+
+// Object to JSON converters.
+//-----------------------------------------------------------------------------
+
+std::string web_sockets::to_json(uint32_t height)
+{
+    boost::property_tree::ptree property_tree;
+    property_tree.put("height", height);
+
+    std::stringstream json_stream;
+    write_json(json_stream, property_tree);
+    return json_stream.str();
+}
+
+std::string web_sockets::to_json(
+    const bc::chain::transaction& transaction)
+{
+    static constexpr auto json_encoding = true;
+
+    const auto property_tree = bc::property_tree(
+        bc::config::transaction(transaction), json_encoding);
+
+    std::stringstream json_stream;
+    write_json(json_stream, property_tree);
+    return json_stream.str();
+}
+
+std::string web_sockets::to_json(const bc::chain::header& header)
+{
+    auto property_tree = bc::property_tree(
+        bc::config::header(header));
+
+    std::stringstream json_stream;
+    write_json(json_stream, property_tree);
+    return json_stream.str();
+}
+
+std::string web_sockets::to_json(const bc::chain::block& block)
+{
+    return to_json(block.header());
+}
+
+std::string web_sockets::to_json(std::string error,
+    const std::error_code& code)
+{
+    boost::property_tree::ptree property_tree;
+
+    property_tree.put("type", "error");
+    property_tree.put("message", error);
```

Comment:
Why not use `code.message()` here and eliminate the `error` parameter, that's what the message method is for.

Reply:
An earlier revision did that, but it doesn't provide any feedback on what specific call caused the error to the client (e.g. if the javascript code issues a series of query commands and one of them fails somewhere).  Either way.

Reply:
If we need more discretion in the error descriptions then I recommend creating a series of web-service specific codes. Let's avoid both the redundant use of the error object and creating a set of loosely correlated error messages.